src/linker/Linker/PreserveSerialization.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Linker.Dataflow;
+using Mono.Linker.Steps;
+
+namespace Mono.Linker
+{
+
+	// This discovers types attributed with certain serialization attributes, to match the old behavior
+	// of xamarin-android. It is not meant to be complete. Unlike xamarin-andorid:
+	// - this will only discover attributed types in marked assemblies
+	// - this will discover types in non-"link" assemblies as well
+	public class PreserveSerialization : BaseSubStep
+	{
+		public override SubStepTargets Targets =>
+			SubStepTargets.Type
+			| SubStepTargets.Field
+			| SubStepTargets.Method
+			| SubStepTargets.Property
+			| SubStepTargets.Event;
+
+		public override void ProcessType (TypeDefinition type)
+		{
+			ProcessAttributeProvider (type);
+		}
+
+		public override void ProcessField (FieldDefinition field)
+		{
+			ProcessAttributeProvider (field);
+		}
+
+		public override void ProcessProperty (PropertyDefinition property)
+		{
+			ProcessAttributeProvider (property);
+		}
+
+		public override void ProcessMethod (MethodDefinition method)
+		{
+			ProcessAttributeProvider (method);
+		}
+
+		public override void ProcessEvent (EventDefinition @event)
+		{
+			ProcessAttributeProvider (@event);
+		}
+
+		void ProcessAttributeProvider (ICustomAttributeProvider provider)
+		{
+			if (!provider.HasCustomAttributes)
+				return;
+
+			var serializedFor = SerializerKind.None;
+
+			foreach (var attribute in provider.CustomAttributes) {
+				if (IsPreservedSerializationAttribute (provider, attribute, out SerializerKind serializerKind))
+					serializedFor |= serializerKind;
+			}
+
+			TypeDefinition type = provider switch {
+				TypeDefinition td => td,
+				FieldDefinition field => field.DeclaringType,
+				PropertyDefinition property => property.DeclaringType,
+				EventDefinition @event => @event.DeclaringType,
+				MethodDefinition method => method.DeclaringType,
+				_ => throw new ArgumentException ($"{nameof (provider)} has invalid provider type {provider.GetType ()}")
+			};
+
+			if (serializedFor.HasFlag (SerializerKind.DataContractSerializer))
+				Context.SerializationHelper.MarkRecursiveMembers (type, new DependencyInfo (DependencyKind.DataContractSerialized, provider));
+			if (serializedFor.HasFlag (SerializerKind.XmlSerializer))
+				Context.SerializationHelper.MarkRecursiveMembers (type, new DependencyInfo (DependencyKind.XmlSerialized, provider));
+		}
+
+		enum SerializerKind
+		{
+			None,
+			XmlSerializer,
+			DataContractSerializer,
+		}
+
+		static bool IsPreservedSerializationAttribute (ICustomAttributeProvider provider, CustomAttribute attribute, out SerializerKind serializerKind)
```

Comment:
I don't think we should use attributes to detect which serializer is needed. We want to trim all these and other serialization-related attributes whenever it's possible.

Reply:
The reason I'm checking for attributes is because we're concerned about back-compat with xamarin-android, and they used to do this. If we're ok with relaxing this, one idea is to make it conditional on marking a serializer ctor.

Reply:
>  If we're ok with relaxing this, one idea is to make it conditional on marking a serializer ctor.

The context back then was different so whatever logic that makes more sense in .NET6 world is preferable.

Reply:
Changed so that we only mark things if we marked a serializer ctor.